### ISSUE TITLE
LDAP group verification doesn't work when using 'dn' as user attribute

### ISSUE THREAD

**Issue by dev-dsp** [gogs/gogs#4684]
- Gogs version (or commit ref):  0.11.28.0718
- Operating system: CentOS 7
- Database:
  - [ ] PostgreSQL
  - [x] MySQL
  - [ ] MSSQL
  - [ ] SQLite
  - [x] Not relevant
- Can you reproduce the bug at https://try.gogs.io:
  - [ ] Yes (provide example URL)
  - [ ] No
  - [x] Not relevant
- Log:
```
2017/08/10 03:17:24 [TRACE] Fetching attributes 'uid', 'givenName', 'sn', 'mail', 'dn' with filter '(uid=testuser)' and base 'uid=testuser,ou=people,dc=example,dc=tk'
2017/08/10 03:17:24 [TRACE] LDAP: Fetching groups 'member' with filter '(cn=gogs-users)' and base 'ou=groups,dc=example,dc=tk'
2017/08/10 03:17:24 [TRACE] LDAP: Group membership test failed [username: testuser, group_member_uid: member, user_uid: 
```

## Description
Group membership verification doesn't work correctly with following options set:
```
Group Attribute Containing List of Users: member
User Attribute Listed in Group: dn
```
'dn' is just not used and 'user_uid' is blank, as you can see in log above
```
$ >>> ldapsearch -x "(cn=gogs-users)" -b 'ou=groups,dc=example,dc=tk' member -LLL
dn: cn=gogs-users,ou=groups,dc=example,dc=tk
member: uid=testuser,ou=people,dc=example,dc=tk
```

**Comment by aboron** [comment]
One bug I see from my pull request a while back is that I introduced a typo on line 272 of the file gogs/pkg/auth/ldap/ldap.go.

this:
`		} else if len(sr.Entries) < 1 {`
should be:
`		} else if len(srg.Entries) < 1 {`

This would affect group membership checking, since it would not be looking at the correct result set.

**Comment by aboron** [comment]
So, it looks like the GetAttributeValue() function will not return a value for an attribute called 'dn' since it considers that to be the identifier of the root object.  So the specific case of searching for group members by 'dn' will have to be tested and handled differently.

Looking into this now.

**Comment by aboron** [comment]
Proposed fix available in develop branch of https://github.com/aboron/gogs - just changed the one [ldap.go](https://raw.githubusercontent.com/aboron/gogs/develop/pkg/auth/ldap/ldap.go) file: 
@dev-dsp - If you could test it I would appreciate that.

**Comment by dev-dsp** [comment]
Hi @aboron,
Is there any way I could patch existing installation, or I need to rebuild it from sources anyway?
I apologize for maybe a stupid question, I have no Go experience :)

**Comment by aboron** [comment]
Go is a compiled language, so it would have to be rebuilt from the source.  I'm so used to doing that I didn't even think about that, sorry.

There are general instructions here at: https://gogs.io/docs/installation/install_from_source.html

Follow that using the "build from develop branch" path, and substitute a git command of:
`git clone --depth=1 -b develop https://github.com/aboron/gogs`
to make my patched version.

If this isn't working out for you, let me know and I might be able to set up a test OpenLdap instance for myself - but I won't be able to get to that for a few days.

**Comment by dev-dsp** [comment]
Ok, thanks a lot. That's not a problem, I'll build it in nearest time and let you know the results.

**Comment by dev-dsp** [comment]
Verified. Confirming as working solution.

**Comment by aboron** [comment]
Pull request sent, thanks for the help debugging.

**Comment by dev-dsp** [comment]
How long it usually takes to approve&merge PR?

**Comment by cheprus** [comment]
Thanks. Waiting impatiently for the review.

**Comment by rickywu** [comment]
I thinks it's the same issue https://github.com/gogits/gogs/issues/4792

**Comment by Unknwon** [comment]
This is claimed to be fixed by merging #4688, please help test on `develop` branch.
